AX: Remove unnecessary AXProperty::SelectedChildren

twilco opened the pull request:
#### 28068289ec6f7379404b5ff52adbf07c18f06bdb
<pre>
AX: Remove unnecessary AXProperty::SelectedChildren
<a href="https://bugs.webkit.org/show_bug.cgi?id=285153">https://bugs.webkit.org/show_bug.cgi?id=285153</a>
<a href="https://rdar.apple.com/142037291">rdar://142037291</a>

Reviewed by Chris Fleizach.

We can compute this property on-demand with information we have already cached off the main-thread. This change allows
us to do less up-front work when creating each object, and guarantees this property can never get stale (because it doesn&apos;t exist).
We also need to do less work when keeping the tree up-to-date (e.g. we no longer need to handle AXSelectedChildrenChanged in updateIsolatedTree).

* LayoutTests/accessibility/combobox/combobox-active-element-selected-children-expected.txt:
* LayoutTests/accessibility/combobox/combobox-active-element-selected-children.html:
* LayoutTests/accessibility/combobox/mac/aria-combobox-activedescendant-expected.txt:
* LayoutTests/accessibility/combobox/mac/aria-combobox-activedescendant.html:
* LayoutTests/accessibility/list/list-active-element-selected-children-expected.txt:
* LayoutTests/accessibility/list/list-active-element-selected-children.html:
* LayoutTests/accessibility/list/listbox-active-element-selected-children-expected.txt:
* LayoutTests/accessibility/list/listbox-active-element-selected-children.html:
* LayoutTests/resources/accessibility-helper.js:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::canHaveSelectedChildren const):
(WebCore::AXCoreObject::selectedChildren):
(WebCore::AXCoreObject::ariaListboxSelectedChildren):
(WebCore::AXCoreObject::ariaSelectedRows):
(WebCore::AXCoreObject::selectedListItems):
(WebCore::AXCoreObject::isActiveDescendantOfFocusedContainer const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AccessibilityListBox.cpp:
(WebCore::AccessibilityListBox::selectedChildren): Deleted.
* Source/WebCore/accessibility/AccessibilityListBox.h:
* Source/WebCore/accessibility/AccessibilityMenuListPopup.cpp:
(WebCore::AccessibilityMenuListPopup::selectedChildren): Deleted.
* Source/WebCore/accessibility/AccessibilityMenuListPopup.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::textAsLabelFor const):
(WebCore::AccessibilityNodeObject::stringValue const):
(WebCore::accessibleNameForNode):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::ariaSelectedRows): Deleted.
(WebCore::AccessibilityObject::ariaListboxSelectedChildren): Deleted.
(WebCore::AccessibilityObject::canHaveSelectedChildren const): Deleted.
(WebCore::AccessibilityObject::selectedChildren): Deleted.
(WebCore::AccessibilityObject::selectedListItems): Deleted.
(WebCore::AccessibilityObject::isActiveDescendantOfFocusedContainer const): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::selectedChildren): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):
* Source/WebCore/inspector/InspectorAuditAccessibilityObject.cpp:
(WebCore::InspectorAuditAccessibilityObject::getSelectedChildNodes):
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::buildObjectForAccessibilityProperties):

Canonical link: <a href="https://commits.webkit.org/288364@main">https://commits.webkit.org/288364@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2ba3a282cac73bc3905ec90d5286c5acbf1e8290

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82813 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2462 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37029 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87936 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33870 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84906 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2534 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10377 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64532 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22294 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85869 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1874 "Found 1 new test failure: fast/forms/ios/file-upload-panel.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75328 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44808 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1766 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29517 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32903 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73021 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30190 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89298 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10104 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7340 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72945 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10332 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71136 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72159 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17920 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16324 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1492 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10057 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15578 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9931 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13396 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11700 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fleizach commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -436,6 +436,151 @@ AXCoreObject* AXCoreObject::selectedTabItem()
+        break;
+    case AccessibilityRole::List: {
+        auto selectedListItems = this->selectedListItems();
+        if (!selectedListItems.isEmpty())
```
Do we need this isEmpty check? I think we'll end up creating an empty vector at the bottom

twilco replied:
Good point! Fixed in the latest commit.

twilco committed:
AX: Remove unnecessary AXProperty::SelectedChildren
https://bugs.webkit.org/show_bug.cgi?id=285153
rdar://142037291

Reviewed by Chris Fleizach.

We can compute this property on-demand with information we have already cached off the main-thread. This change allows
us to do less up-front work when creating each object, and guarantees this property can never get stale (because it doesn't exist).
We also need to do less work when keeping the tree up-to-date (e.g. we no longer need to handle AXSelectedChildrenChanged in updateIsolatedTree).

* LayoutTests/accessibility/combobox/combobox-active-element-selected-children-expected.txt:
* LayoutTests/accessibility/combobox/combobox-active-element-selected-children.html:
* LayoutTests/accessibility/combobox/mac/aria-combobox-activedescendant-expected.txt:
* LayoutTests/accessibility/combobox/mac/aria-combobox-activedescendant.html:
* LayoutTests/accessibility/list/list-active-element-selected-children-expected.txt:
* LayoutTests/accessibility/list/list-active-element-selected-children.html:
* LayoutTests/accessibility/list/listbox-active-element-selected-children-expected.txt:
* LayoutTests/accessibility/list/listbox-active-element-selected-children.html:
* LayoutTests/resources/accessibility-helper.js:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::canHaveSelectedChildren const):
(WebCore::AXCoreObject::selectedChildren):
(WebCore::AXCoreObject::ariaListboxSelectedChildren):
(WebCore::AXCoreObject::ariaSelectedRows):
(WebCore::AXCoreObject::selectedListItems):
(WebCore::AXCoreObject::isActiveDescendantOfFocusedContainer const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AccessibilityListBox.cpp:
(WebCore::AccessibilityListBox::selectedChildren): Deleted.
* Source/WebCore/accessibility/AccessibilityListBox.h:
* Source/WebCore/accessibility/AccessibilityMenuListPopup.cpp:
(WebCore::AccessibilityMenuListPopup::selectedChildren): Deleted.
* Source/WebCore/accessibility/AccessibilityMenuListPopup.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::textAsLabelFor const):
(WebCore::AccessibilityNodeObject::stringValue const):
(WebCore::accessibleNameForNode):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::ariaSelectedRows): Deleted.
(WebCore::AccessibilityObject::ariaListboxSelectedChildren): Deleted.
(WebCore::AccessibilityObject::canHaveSelectedChildren const): Deleted.
(WebCore::AccessibilityObject::selectedChildren): Deleted.
(WebCore::AccessibilityObject::selectedListItems): Deleted.
(WebCore::AccessibilityObject::isActiveDescendantOfFocusedContainer const): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::selectedChildren): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):
* Source/WebCore/inspector/InspectorAuditAccessibilityObject.cpp:
(WebCore::InspectorAuditAccessibilityObject::getSelectedChildNodes):
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::buildObjectForAccessibilityProperties):

Canonical link: https://commits.webkit.org/288364@main
